src/controllers/validators/PenaltyDetailsValidator.ts

```diff
@@ -0,0 +1,93 @@
+import { AnySchema } from '@hapi/joi';
+import { SessionKey } from 'ch-node-session-handler/lib/session/keys/SessionKey';
+import { ISignInInfo } from 'ch-node-session-handler/lib/session/model/SessionInterfaces';
+import { Penalty, PenaltyList } from 'ch-sdk-node/dist/services/lfp/types';
+import Resource from 'ch-sdk-node/dist/services/resource';
+import { Request } from 'express';
+import { OK } from 'http-status-codes';
+
+import { Validator } from 'app/controllers/validators/Validator';
+import { loggerInstance } from 'app/middleware/Logger';
+import { PenaltyIdentifier } from 'app/models/PenaltyIdentifier';
+import { CompaniesHouseSDK, OAuth2 } from 'app/modules/Types';
+import { getEnvOrThrow } from 'app/utils/EnvironmentUtils';
+import { SchemaValidator } from 'app/utils/validation/SchemaValidator';
+import { ValidationError } from 'app/utils/validation/ValidationError';
+import { ValidationResult } from 'app/utils/validation/ValidationResult';
+
+export class PenaltyDetailsValidator implements Validator {
+    constructor(private readonly formSchema: AnySchema,
+                private readonly chSdk: CompaniesHouseSDK) {}
+
+
+    private createValidationResultWithErrors(): ValidationResult {
+        return new ValidationResult([
+            new ValidationError('companyNumber', 'Check that you’ve entered the correct company number'),
+            new ValidationError('penaltyReference', 'Check that you’ve entered the correct reference number')
+        ]);
+    }
+
+    async validate(request: Request): Promise<ValidationResult> {
+        const API_URL = getEnvOrThrow(`APPEALS_API_URL`);
+
+        const schemaResults: ValidationResult = new SchemaValidator(this.formSchema).validate(request.body);
+
+        if (schemaResults.errors.length > 0) {
+            return schemaResults;
+        }
+
+        if (!request.session) {
+            throw new Error('Session is undefined');
+        }
+
+        const signInInfo: ISignInInfo | undefined = request.session!.get<ISignInInfo>(SessionKey.SignInInfo);
+
+        const accessToken: string | undefined = signInInfo?.access_token?.access_token;
+
+        const companyNumber: string = (request.body as PenaltyIdentifier).companyNumber;
+
+        const penaltyReference: string = (request.body as PenaltyIdentifier).penaltyReference;
+
+        try {
+
+            const penalties: Resource<PenaltyList> =
+                await this.chSdk(new OAuth2(accessToken!)).lateFilingPenalties.getPenalties(companyNumber);
+
+            if (penalties.httpStatusCode !== OK) {
+                throw new Error(`AppealDetailActionProcessor: failed to get penalties from pay API with status code ${penalties.httpStatusCode} with access token ${accessToken} and base url ${API_URL}`);
+            }
+
+            let items: Penalty[];
```

Comment:
Initialise items to [] so you don't have to check if it's undefined later

Reply:
you could shorten this to:
```ts
let items: Penalty[] = penalties.resource.items.filter(penalty => penalty.type === 'penalty');
if (modernPenaltyReferenceRegex.test(penaltyReference)){
    items = items.filter(penalty.id === penaltyReference.substring(1));
}
```

Reply:
Good one